Source/Core/CoverageAnnotator.cs

```diff
@@ -0,0 +1,115 @@
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.Boogie;
+
+/// <summary>
+/// Add `{:id ...}` attributes to all assertions, assumptions, requires
+/// clauses, ensures clauses, and call statements so that verification
+/// coverage tracking is possible. This exists primarily to support the
+/// automatic detection of vacuous proofs in the case where no front
+/// end has added these already.
+/// </summary>
+public class CoverageAnnotator : StandardVisitor
+{
+  private int idCount = 0;
+  private string currentImplementation;
+  private Dictionary<string, ISet<string>> implementationGoalIds = new();
+  private Dictionary<string, Absy> idMap = new();
+
+  private void AddImplementationGoalId(string id)
+  {
+    implementationGoalIds[currentImplementation].Add(id);
+  }
+  
+  private void AddIdIfMissing(ICarriesAttributes node, bool isGoal)
+  {
+    Absy absy = node as Absy;
+    if (absy == null) {
+      return;
+    }
+    var idStr = node.FindStringAttribute("id");
+    if (idStr == null) {
+      idCount++;
+      idStr = $"id_l{absy.tok.line}_c{absy.tok.col}_{NodeType(node)}_{idCount}";
```

Comment:
```suggestion
      idStr = $"id_l{absy.tok.line}_c{absy.tok.col}_{NodeType(node)}_{idCount}";
      idCount++;
```
We don't want to miss out on `0`.

Reply:
Yeah, poor, lonely zero. It'd be even lonelier than one.